chain/cardano/client/client.go

```diff
@@ -0,0 +1,407 @@
+)
+
+// Client for Template
+type Client struct {
+	Asset               xc.ITask
+	Url                 string
+	Network             string
+	Logger              *log.Entry
+	BlockfrostProjectId string
+	HttpClient          *http.Client
+}
+
+var _ xclient.Client = &Client{}
+
+// NewClient returns a new Template Client
+func NewClient(cfgI xc.ITask) (*Client, error) {
+	cfg := cfgI.GetChain()
+	url := cfg.GetChain().URL
+	if url == "" {
+		return nil, errors.New("rpc url is empty")
+	}
+
+	network := cfg.GetChain().Network
+	if network == "" {
+		network = "mainnet"
+		log.Warn("network is empty, defaulting to mainnet")
+	}
+
+	logger := log.WithFields(log.Fields{
+		"chain":   cfg.Chain,
+		"rpc":     url,
+		"network": network,
+	})
+
+	return &Client{
+		Asset:               cfgI,
+		Url:                 url,
+		Network:             network,
+		Logger:              logger,
+		BlockfrostProjectId: cfg.Auth2.LoadOrBlank(),
+		HttpClient:          http.DefaultClient,
+	}, nil
+}
+
+func (client *Client) FetchProtocolParameters(ctx context.Context) (types.ProtocolParameters, error) {
+	var protocolParameters types.ProtocolParameters
+	err := client.Request(ctx, GET, "/epochs/latest/parameters", nil, &protocolParameters)
+
+	return protocolParameters, err
+}
+
+func (client *Client) FetchUtxos(ctx context.Context, address xc.Address, contract xc.ContractAddress) ([]types.Utxo, error) {
+	path := fmt.Sprintf("/addresses/%s/utxos/%s?order=desc", string(address), string(contract))
+	var response []types.Utxo
+	err := client.Request(ctx, GET, path, nil, &response)
+	return response, err
+}
+
+// 1. Sort utxos by amount descending
+// 2. Get minimum utxo set that adds up to the target amount
+func GetMinUtxoSet(utxos []types.Utxo, targetAmount xc.AmountBlockchain, contract xc.ContractAddress) []types.Utxo {
+	slices.SortFunc(utxos, func(lhs types.Utxo, rhs types.Utxo) int {
+		amountL := lhs.GetAssetAmount(contract)
+		amountR := rhs.GetAssetAmount(contract)
+
+		return amountR.Cmp(&amountL)
+	})
+
+	utxoSet := make([]types.Utxo, 0)
+	balance := xc.NewAmountBlockchainFromUint64(0)
+	for _, utxo := range utxos {
+		if balance.Cmp(&targetAmount) >= 0 {
+			break
+		}
+
+		amount := utxo.GetAssetAmount(contract)
+		balance = balance.Add(&amount)
+		utxoSet = append(utxoSet, utxo)
+	}
+	return utxoSet
+}
+
+// FetchTransferInput returns tx input for a Cardano transfer
+func (client *Client) FetchTransferInput(ctx context.Context, args xcbuilder.TransferArgs) (xc.TxInput, error) {
+	decimals, err := client.FetchDecimals(ctx, Lovelace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch decimals: %w", err)
+	}
+
+	contract, ok := args.GetContract()
+	if !ok {
+		contract = Lovelace
+	}
+
+	utxos, err := client.FetchUtxos(ctx, args.GetFrom(), contract)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch utxos: %w", err)
+	}
+
+	// Include the fee in target amount
+	cfg := client.Asset.GetChain()
+	feeLimit := cfg.FeeLimit
```

Comment:
I'd map `GasBudgetDefault` instead.

* `GasBudgetDefault` is one of the chain _client_ configurations, which is easy to change.
* `FeeLimit` is chain _base_ configuration, meant for the user/custodial solution to use as a safety limit protect from RPC fee griefing.